src/atom.rs

```diff
@@ -0,0 +1,166 @@
+            (Underscore(l), Underscore(r)) => l.cmp(r),
+        }
+    }
+}
+
+fn cmp_ignore_case(l: &str, r: &str) -> Ordering {
+    for (a, b) in l.bytes().zip(r.bytes()) {
+        match a.to_ascii_lowercase().cmp(&b.to_ascii_lowercase()) {
+            Ordering::Equal => match a.cmp(&b) {
+                Ordering::Equal => {}
+                non_eq => return non_eq,
+            },
+            non_eq => return non_eq,
+        }
+    }
+
+    l.len().cmp(&r.len())
+}
+
+pub fn iter_atoms(string: &str) -> AtomIter {
+    AtomIter {
+        bytes: string.as_bytes(),
+        offset: 0,
+    }
+}
+
+pub struct AtomIter<'a> {
+    bytes: &'a [u8],
+    offset: usize,
+}
+
+// https://tools.ietf.org/html/rfc3629
+#[cfg_attr(rustfmt, rustfmt_skip)]
+static UTF8_CHAR_WIDTH: [u8; 256] = [
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // 0x1F
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // 0x3F
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // 0x5F
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,
+    1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1, // 0x7F
+    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, // 0x9F
+    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,
+    0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0, // 0xBF
+    0,0,2,2,2,2,2,2,2,2,2,2,2,2,2,2,
+    2,2,2,2,2,2,2,2,2,2,2,2,2,2,2,2, // 0xDF
+    3,3,3,3,3,3,3,3,3,3,3,3,3,3,3,3, // 0xEF
+    4,4,4,4,4,0,0,0,0,0,0,0,0,0,0,0, // 0xFF
+];
+
+impl<'a> Iterator for AtomIter<'a> {
+    type Item = Atom<'a>;
+
+    fn next(&mut self) -> Option<Atom<'a>> {
+        // No more characters.
+        if self.offset >= self.bytes.len() {
+            return None;
+        }
+
+        let x = self.bytes[self.offset];
+
+        // Underscore or Number.
+        match x {
+            // Underscore.
+            b'_' => {
+                self.offset += 1;
+
+                let mut n = 1;
+                while self.offset < self.bytes.len() {
+                    match self.bytes[self.offset] {
+                        b'_' => {
+                            self.offset += 1;
+                            n += 1;
+                        }
+                        _ => break,
+                    }
+                }
+
+                Some(Atom::Underscore(n))
+            }
+            // Number.
+            b'0'..=b'9' => {
+                let start = self.offset;
+
+                self.offset += 1;
+                while self.offset < self.bytes.len() {
+                    match self.bytes[self.offset] {
+                        b'0'..=b'9' => self.offset += 1,
+                        _ => break,
+                    }
+                }
+
+                // This is safe since we've verified it's valid utf8.
+                let s = &self.bytes[start..self.offset];
+                let s = unsafe { from_utf8_unchecked(s) };
+
+                let int = match s.parse::<u64>() {
```

Comment:
This shouldn't be required, and crashes if a number is big.

```console
error: custom attribute panicked
  --> tests/order.rs:30:1
   |
30 | #[remain::sorted]
   | ^^^^^^^^^^^^^^^^^
   |
   = help: message: internal error: entered unreachable code: expected integer
```

There is a simpler way to compare base 10 numbers without converting to base 2: first by length then by ordinary alphanumeric comparison.